### ISSUE TITLE
noAccess error result for inaccessible OIDs in Get, GetNext, GetBulk

### ISSUE THREAD

**Issue by u8675309** [markabrahams/node-net-snmp#179]
I have a MIB with table indexes set as 'accessible-for-notify'. This would mean they cannot be obtained with Get, GetNext, and GetBulk. Looking at RFC-3416, it looks like Get for an inaccessible OID should respond with "noSuchInstance" and that GetNext and GetBulk should skip over the inaccessible OID (since they process the orderd list of accessible OIDs).

What happens instead is that a "noAccess" response is sent. I believe "noAccess" should only be used for a Set request (and Get, GetNext, GetBulk should act as if it does not exist).

The result of this issue is that snmpwalk and snmpgetbulk do not retrieve the expected records.

### How to run agent
```$ npm install net-snmp```
```$ node .```

### How to run SNMP commands
```snmpwalk -v2c -c public localhost:1161 1```
```snmpgetnext -v2c -c public localhost:1161 1```
```snmpget -v2c -c public localhost:1161 1.3.6.1.4.1.534.6.6.7.7.3.1.1.1```

### Actual behavior
```
$ snmpwalk -v2c -c public localhost:1161 1
Error in packet.
Reason: noAccess
Failed object: SNMPv2-SMI::enterprises.534.6.6.7.7.3.1.1.1

SNMPv2-SMI::internet = No Such Object available on this agent at this OID
$ snmpgetnext -v2c -c public localhost:1161 1
Error in packet.
Reason: noAccess
Failed object: SNMPv2-SMI::enterprises.534.6.6.7.7.3.1.1.1

$ snmpget -v2c -c public localhost:1161 1.3.6.1.4.1.534.6.6.7.7.3.1.1.1
Error in packet
Reason: noAccess
Failed object: SNMPv2-SMI::enterprises.534.6.6.7.7.3.1.1.1
```

### Expected behavior
```
$ snmpwalk -v2c -c public localhost:1161 1
SNMPv2-SMI::enterprises.534.6.6.7.7.3.1.2.1 = STRING: "Some Name"
SNMPv2-SMI::enterprises.534.6.6.7.7.3.1.2.1 = No more variables left in this MIB View (It is past the end of the MIB tree)
$ snmpgetnext -v2c -c public localhost:1161 1
SNMPv2-SMI::enterprises.534.6.6.7.7.3.1.2.1 = STRING: "Some Name"
$ snmpget -v2c -c public localhost:1161 1.3.6.1.4.1.534.6.6.7.7.3.1.1.1
SNMPv2-SMI::enterprises.534.6.6.7.7.3.1.1.1 = No Such Object available on this agent at this OID
```

[demo.zip](https://github.com/markabrahams/node-net-snmp/files/6239332/demo.zip)

**Comment by u8675309** [comment]
I think the update for GetNext and GetBulk would involve calling addGetNextVarbind in a loop as long as the returned nodes are not accessible. This is called from agent.getNextRequest and agent.getBulkRequest.

I think the update for Get would be changing agent.request to send "NoSuchObject" instead of "noAccess" for Get requests.

**Comment by u8675309** [comment]
Or possibly just changing Mib.prototype.addTableRow to not add "accessible-for-notify" column nodes (in addition to not adding them for "not-accessible"). I'm not sure if that would cause any problems elsewhere.

**Comment by markabrahams** [comment]
Hi @u8675309 - yes, I agree with all of your observations above.  Returning "noAccess" seems wrong in this case; "noSuchObject" is a better return.  From RFC 3416 Section 4.1.2, point (2):

> (2)   Otherwise, if the variable binding's name does not have an
>          OBJECT IDENTIFIER prefix which exactly matches the OBJECT
>          IDENTIFIER prefix of any (potential) variable accessible by
>          this request, then its value field is set to "noSuchObject".

However, the suggestion in your last post is probably the way I would steer.  We went this way for "not-accessible" based on evidence that Net-SNMP was doing this, given that there was no solid RFC guidance on how to handle this scenario - see discussion in #145 and in particular this post:
https://github.com/markabrahams/node-net-snmp/issues/145#issuecomment-753683372

So in the absence of RFC guidance, I'm left with:
(1) Following any example of how the Net-SNMP agent handles "accessible-for-notify" index components.
(2) Failing to find such evidence, I'd suggest being somewhat consistent with what we've done for "not-accessible" in `addTableRow()` i.e. not adding the column nodes as you say.  To answer your last question, this shouldn't cause any problems elsewhere, as it's already a code path that's been successfully implemented for "not-accessible" index components.

If you're happy to research (1) and/or provide a pull request for (2) - or I could do (2) when I get a moment - either would be an acceptable solution for mind.

**Comment by markabrahams** [comment]
Hi @u8675309 - I've implemented solution (2) above and published the change in version 3.5.4 of the npm.
